### PR TITLE
Implement Variable.new

### PR DESCRIPTION
**Pull request by colesbury**
This implements Variable.new(). We'll need to modify it when we make the `torch.Tensor` constructors return the combined Variable/Tensor type, but it should be a useful starting point for that too.

I've been putting a bunch of Tensor/Variable functions in `torch/csrc/utils` (tensor_new, tensor_list, tensor_numpy). Once we merge Variable and Tensor we should find a better place for them to live.

cc @zdevito this adds `unsafeStorageFromTH` to `at::Type`

### CODE REVIEW

File: torch/csrc/utils/python_arg_parser.cpp
Comment (gchanan) [inline torch/csrc/utils/python_arg_parser.cpp]: do we need this?  We had parameters successfully taking sizes before, right (even if they were specified as IntLists in the parsers).

File: torch/csrc/utils/python_arg_parser.cpp
Comment (gchanan) [inline torch/csrc/utils/python_arg_parser.cpp]: do we still need deprecated?  I don't see it being read in python_arg_parser anymore.

File: torch/csrc/utils/tensor_list.h
Comment (gchanan) [inline torch/csrc/utils/tensor_list.h]: is this supposed to be part of a general pattern going forward?  i.e. prefer naming torch::utils::name rather than torch::THPUtils_name?

File: torch/csrc/utils/tensor_new.cpp
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: When we have scalars, it won't be great that var.new(5) is equivalent to var.new([5]), not numpy.array(5).  This function already seems overly complex before we even get there -- it seems like we should split it up.  IMO there are really two functions here: size-based and data-based factories (the first two of parser1 and the only of parser2 are size-based, everything else is data-based).  I don't really like these names, but the best I could come up with at the moment are "new" for the size based" and "new_init" for the data-based (i.e. it is equivalent to new + initialization).  I'm sure we can come up with something better.

**Review by gchanan — CHANGES_REQUESTED** [review]
This looks generally good, but I don't think we should overload "new" with size and data semantics (particularly in the case of scalars).  It is IMO confusing to the user and also complicates the code (e.g. the need for separate torch.Size parsing).

File: aten/src/ATen/UndefinedType.h
Comment (colesbury) [reply aten/src/ATen/UndefinedType.h]: I think we should just get rid of Storage and replace most uses with a 1-D tensor. But that should wait until after we merge Tensor and Variable.

File: test/test_torch.py
Comment (colesbury) [reply test/test_torch.py]: It is the same semantics as `Tensor.new`. `x.new()` returns a new empty tensor.

We've been returning `[0]` for the shape since https://github.com/pytorch/pytorch/commit/067f799e9ff3cc24010492d512ad610caf4143ea (three weeks ago).

File: test/test_torch.py
Comment (colesbury) [reply test/test_torch.py]: It's not easy right now. It'll be easier after we merge Tensor and Variable and only have deal with `VariableType` in C++.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (colesbury) [reply torch/csrc/utils/python_arg_parser.cpp]: `torch.Size()` really only exists for the tensor constructors to disambiguate size vs. data constructors. This binds to `torch.Size` but not `tuple`. We could remove this from here and do the disambiguation after the parsing in `tensor_new.cpp`, but we still need to handle it specially.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (colesbury) [reply torch/csrc/utils/python_arg_parser.cpp]: Yeah. We don't currently issue warning, but we will after we merge tensor and variable.

File: torch/csrc/utils/tensor_list.h
Comment (colesbury) [reply torch/csrc/utils/tensor_list.h]: Yes, I think we should use C++ namespaces instead of name prefixes like `TH`/`THP`. I think it's also good practice to have the namespace correspond to the file path.

File: torch/csrc/utils/tensor_new.cpp
Comment (colesbury) [reply torch/csrc/utils/tensor_new.cpp]: We need to match the semantics of `Tensor.new` to avoid breaking backwards compatibility. We can deprecate some of the overloads, but they all need to be implemented.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (gchanan) [reply torch/csrc/utils/python_arg_parser.cpp]: I see. I think we should do the disambiguation after the parsing because it's really a special case.  I could see people writing their own methods and specifying torch.Size when they really mean IntList if we keep this.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (gchanan) [reply torch/csrc/utils/python_arg_parser.cpp]: What's the reason for not issuing a warning now?

File: torch/csrc/utils/tensor_new.cpp
Comment (gchanan) [reply torch/csrc/utils/tensor_new.cpp]: deprecating some of the overloads sounds like a good solution.  What do you think about having separate size vs data factories?  Which one should "new" be?

File: torch/csrc/utils/python_arg_parser.cpp
Comment (colesbury) [reply torch/csrc/utils/python_arg_parser.cpp]: Some of the new signatures are only implemented on Variable, so the "correct" way to call (e.g. baddbmm) on Variables and Tensors would differ. I'd rather wait until after we kill Tensor to start issuing warnings.

File: torch/csrc/utils/tensor_new.cpp
Comment (colesbury) [reply torch/csrc/utils/tensor_new.cpp]: I think `torch.Tensor()` and `Tensor.new()` should be the size-based constructors.
`torch.tensor()` and (maybe `Tensor.tensor`) should be the data-based factories.

This more-or-less matches numpy where the type `numpy.ndarray` takes a shape and `numpy.array` takes data.

I'd like to get rid of the the other constructors:
`torch.Tensor(tensor)` is equivalent to `tensor[:]`
`torch.Tensor(_cdata=...)` could be `torch._unsafe_from_cdata` or something. (I'm not even sure we use it anymore.
`torch.Tensor(storage)` could be `torch.from_storage(storage)` or we could get rid of storages from the Python API.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (gchanan) [inline torch/csrc/utils/python_arg_parser.cpp]: which case does this fix?

**Review by gchanan — APPROVED** [review]
lgtm.

File: torch/csrc/utils/python_arg_parser.cpp
Comment (colesbury) [reply torch/csrc/utils/python_arg_parser.cpp]: We we're incorrectly matching some signatures because we decremented `remaining_kwargs` even when there wasn't a kwarg. For example, from the tests:

```
x.new(sizes=(1, 2, 3))
```

Was matching:

```
Tensor(*, int64_t device=-1)
```

because device is optional and we incorrectly decremented remaining_kwargs to 0 after we set `device` to it's default value (-1).

**Pull request merged**
